cv verification result topics (#1901)
harness/developer-hub: docs/continuous-delivery/verify/cv-results/interpret-log-results.md

---
title: Interpret log verification results
sidebar_position: 3
description: Learn how to interpret log verification results. 
---

Logs emitted by a service provide valuable insights into its operations. Developers typically use logs for the following purposes:

> - Support custom metrics: Logs contain information that can be used to aggregate custom metrics.

Reviewer: ```suggestion
- Support custom metrics — Contains information that can be used to aggregate custom metrics.
```

Author: Updated

> - Debugging: Logs help developers diagnose and troubleshoot issues reported in the service.

Reviewer: Accept as you see fit. If it is not consistent, please ignore it.

> - Debugging: Logs help developers diagnose and troubleshoot issues reported in the service.

Author: Updated

During the debugging process, developers often perform the following tasks:

- Check for new errors: They examine if any new errors have been reported by the service.

- Monitor error frequency: They observe whether the occurrence frequency of any errors has changed.

Doing these tasks manually can be cumbersome. Developers usually start by filtering errors one by one until they come across new ones. They might use wildcard-based filters to exclude similar errors. However, there is no fixed pattern for these filters, as they vary depending on the specific errors.  

Now, let's consider the context of an issue reported after a recent deployment. Developers manually analyze the logs to identify new errors and changes in error frequency compared to the previous version of the service.

This is where log analysis with Harness Continuous Verification (CV) comes into play for developers. It offers the following benefits:

- Log analysis helps automatically identify new errors and changes in error frequency after deployment before they become widespread issues.

- Debugging the issue.

> Logging tools such as Google Cloud Operations (GCP) and Splunk help you capture and analyze logs for various purposes, such as debugging, monitoring, auditing, and security. To see the list of logging tools supported by Harness, go to [Log providers](https://developer.harness.io/docs/getting-started/supported-platforms-and-technologies#log-providers).

Reviewer: ```suggestion
Logging tools such as Google Cloud Platform (GCP) and Splunk help you capture and analyze logs for various purposes, such as debugging, monitoring, auditing, and security. To see the list of logging tools supported by Harness, go to [Log providers](https://developer.harness.io/docs/getting-started/supported-platforms-and-technologies#log-providers).
```

Author: Removed GCP and retained Google Cloud Operations

## Log verification in Harness CV

You can [configure your logging tool](/docs/category/configure-cv) (also known as a health source) in Harness CV. Harness CV uses the logs provided by your logging tool to analyze errors. Harness CV log analysis employs a clustering mechanism to group similar errors together and compares them with error clusters from previous service versions. Harness CV uses Machine Learning (ML) in the clustering and comparison process because it cannot be accomplished using fixed user-defined rules alone. For more information on how Harness CV uses ML, go to Machine Learning Usage in Continuous Verification.

Harness CV classifies the error clusters identified by the ML model during log analysis into the following categories:

- Known: Errors that were also present in the previous version.

- Unknown: New errors that have not been observed before.

> - Unexpected Frequency: Errors with an increased occurrence frequency.

Reviewer: Are these UI elements? If so, we might have to make them bold-faced. Otherwise, we don't need to capitalize F in frequency, right?

Author: Updated

## View log verification results

Harness CV provides a powerful mechanism to visualize and interpret the log verification results.

> Prerequisite: You should have added a health source to the Verify step and run the pipeline. To learn more about how to add a health source to the verify step and run the pipeline, go to Configure CV.

Reviewer: ```suggestion
**Prerequisite**: You should have added a health source to the Verify step and run the pipeline. To learn more about how to add a health source to the verify step and run the pipeline, go to Configure CV.
```

Reviewer: A hyperlink to Configure CV?

Author: Updated

To view log verification results for a pipeline:

1. In your Harness project, navigate to the **Deployments** module > **Pipelines**, and then select a pipeline for which you want to view the verification results.

2. In the top right corner, select **Execution History**, and from the pipeline executions list, choose the execution for which you want to see the verification results.

> 3. Ensure that the **Console View** toggle switch is turned ON.

Reviewer: ```suggestion
3. Ensure that the **Console View** toggle switch is turned on.
```

Author: Updated

4. From the stages list, expand the stage in which you want to see the verification results, and then select **Verify**.  
   Verification details appear on the right.

5. Go to the **Logs** tab to see the log verification details.

The Logs tab displays the following details:

- A Radar Chart

- Event list

<docimage path={require('./static/log-verification.png')} />

## Radar Chart

The Radar chart presents a comprehensive bird's-eye view of all the log clusters detected during log verification. This visual representation allows you to quickly understand the distribution and magnitude of each type of cluster. 

<docimage path={require('./static/radar-chart.png')} />

> The radar chart visually organizes log clusters into three concentric circles. The outermost circle represents all the Unknown event clusters, while the middle circle contains both Known and Unexpected frequency clusters.

Reviewer: Do we need to capitalize Unknown, Known, etc?

Author: Updated

:::

### Event health

Event health has three possible values: **Healthy**, **Medium Healthy**, and **Unhealthy**. These values are directly associated with the Known, Unexpected frequency, and Unknown event types, respectively, if you have not set any event preferences. However, you have the ability to modify this mapping by setting preferences for a log cluster. For example, marking an "unknown" cluster as non-risky would change its classification to "Healthy." For more information on the event preference setting feature, go to [Event preference or log feedback](./log-feedback.md).

### Sensitivity and event health 

The event health plays a crucial role in the verification process and is determined based on the sensitivity configuration of the verify step. The rules for determining verification outcomes are as follows:

- High sensitivity: Verification fails if there is any Unhealthy or Medium Healthy cluster.

- Medium sensitivity: Verification fails if there is any Unhealthy cluster.

- Low sensitivity: Verification fails if there is any Unhealthy cluster.

> 

Reviewer: If the description for both Medium and Low are the same, then what's the difference between the two?

Author: Good point. I have floated this question to the dev team. Will make the changes (via another PR) when they respond.
